app/frontend/terrier/parts/select-field-part.ts

```diff
@@ -0,0 +1,214 @@
+     * Assumes that all child elements have the same height.
+     * @param dropdown the element to reposition
+     * @param selectField the anchor select field
+     * @param selectedElement the selected option that will be positioned directly above the select field
+     */
+    anchorSelectDropdown(dropdown: HTMLElement, selectField: HTMLElement, selectedElement: HTMLElement) {
+        const dropdownSize = Overlays.getElementSize(dropdown)
+        const selectOptionSize = Overlays.getElementSize(selectedElement)
+
+        const selectedOptionIndex = Array.prototype.indexOf.call(dropdown.childNodes, selectedElement)
+
+        const anchorRect = selectField.getBoundingClientRect()
+        const win = Overlays.getWindowSize()
+
+        const result: AnchorResult = {
+            top: anchorRect.y - (selectOptionSize.height * selectedOptionIndex),
+            left: anchorRect.x,
+            valid: true
+        }
+
+        Overlays.clampAnchorResult(result, dropdownSize, win)
+
+        if (dropdownSize.width < anchorRect.width) {
+            result.width = anchorRect.width
+        }
+
+        // how far down the select options list we need to scroll
+        let scrollAmount = (selectOptionSize.height * (selectedOptionIndex))
+        const hasOptGroups = this.state.options.find((o) => 'group' in o)
+        if (hasOptGroups) {
+            scrollAmount -= selectOptionSize.height
+        }
+
+        let scrollTop = scrollAmount - anchorRect.top
+        let dropdownIsShrunk = false
+
+        if (dropdownSize.height > win.height - anchorRect.top) { // attempting to fill the whole window height with the dropdown, so we may need to adjust placement of dropdown
+            if (scrollAmount < anchorRect.top) { // need space above the dropdown
+                result.top = anchorRect.top - scrollAmount
+                result.height = win.height - result.top
+                scrollTop = 0
+                dropdownIsShrunk = true
+            } else if (dropdownSize.height - scrollAmount < win.height) { // need space below the dropdown
+                result.height = dropdownSize.height - scrollAmount + anchorRect.top
+                scrollTop = scrollAmount
+                dropdownIsShrunk = true
+            }
+        } else {
+            if (hasOptGroups) {
+                result.top += selectOptionSize.height
+            }
+        }
+
+        let styleString = ""
+        for (const key of ['top', 'left', 'width', 'height'] as const) {
+            if (result[key] != null) {
+                styleString += `${key}: ${result[key]}px;`
+            }
+        }
+
+        dropdown.setAttribute('style', styleString)
+        dropdown.scrollTo({ top: scrollTop })
+
+        dropdown.classList.add('show')
+        selectedElement.classList.add('hover')
+        selectedElement.addEventListener('mouseleave', () => selectedElement.classList.remove('hover'))
```

Comment:
Won't the selected element be under the user's mouse by default anyway?

Reply:
without this class, the 'hover' effect doesn't appear until you move the mouse. I looked into it and was finding that there isn't a way to manually force the ':hover' state, so I went with adding a class instead.